sn_node/tests/spend_simulation.rs

```diff
@@ -0,0 +1,697 @@
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+mod common;
+use assert_fs::TempDir;
+use assert_matches::assert_matches;
+use common::client::{get_client_and_funded_wallet, get_wallet};
+use eyre::{bail, OptionExt, Report, Result};
+use itertools::Itertools;
+use rand::{seq::IteratorRandom, Rng};
+use sn_client::Client;
+use sn_logging::LogBuilder;
+use sn_networking::NetworkError;
+use sn_transfers::{
+    rng, CashNote, DerivationIndex, HotWallet, MainPubkey, NanoTokens, OfflineTransfer,
+    SpendAddress, SpendReason, Transaction, UniquePubkey,
+};
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    path::PathBuf,
+    time::Duration,
+};
+use tokio::sync::mpsc;
+use tracing::*;
+
+const MAX_WALLETS: usize = 50;
+const MAX_CYCLES: usize = 10;
+const AMOUNT_PER_RECIPIENT: NanoTokens = NanoTokens::from(1000);
+/// The chance for an attack to happen. 1 in X chance.
+/// The attack can be one of these:
+/// 1. A double spend of a transaction whose outputs are partially spent / partially UTXO
+/// 2. A double spend of an entire transaction
+/// 3. Poisoning of a transaction whose outputs are all spent.
+const ONE_IN_X_CHANCE_FOR_AN_ATTACK: u32 = 2;
+
+enum WalletAction {
+    Send {
+        recipients: Vec<(NanoTokens, MainPubkey, DerivationIndex)>,
+    },
+    DoubleSpend {
+        cashnotes: Vec<CashNote>,
+        to: (NanoTokens, MainPubkey, DerivationIndex),
+    },
+    ReceiveCashNotes {
+        from: usize,
+        cashnotes: Vec<CashNote>,
+    },
+}
+
+enum WalletTaskResult {
+    Error {
+        id: usize,
+        err: String,
+    },
+    DoubleSpendSuccess {
+        id: usize,
+    },
+    SendSuccess {
+        id: usize,
+        recipient_cash_notes: Vec<CashNote>,
+        change_cash_note: Option<CashNote>,
+        transaction: Transaction,
+    },
+    ReceiveSuccess {
+        id: usize,
+        received_cash_note: Vec<CashNote>,
+    },
+}
+
+#[derive(Debug)]
+enum SpendStatus {
+    Utxo,
+    Spent,
+    Poisoned,
+    // DoubleSpend,
+}
+
+#[derive(custom_debug::Debug)]
+struct State {
+    // immutable
+    #[debug(skip)]
+    action_senders: BTreeMap<usize, mpsc::Sender<WalletAction>>,
+    all_wallets: BTreeMap<usize, TempDir>,
+    main_pubkeys: BTreeMap<usize, MainPubkey>,
+    main_pubkeys_inverse: BTreeMap<MainPubkey, usize>,
+    // mut
+    cashnote_tracker: BTreeMap<UniquePubkey, (SpendStatus, CashNote)>,
+    cashnotes_per_wallet: BTreeMap<usize, Vec<UniquePubkey>>,
+    outbound_transactions_per_wallet: BTreeMap<usize, BTreeSet<Transaction>>,
+}
+
+#[derive(Debug, Default)]
+struct PendingTasksTracker {
+    pending_send_results: Vec<usize>,
+    pending_receive_results: Vec<usize>,
+}
+
+#[tokio::test]
+async fn cash_note_transfer_double_spend_fail() -> Result<()> {
```

Comment:
may need a more detailed comment to explain: what's the detailed scenary of the test, and the expected result .
some general description shall be enough.